sdk/eventhub/Microsoft.Azure.WebJobs.Extensions.EventHubs/src/Listeners/EventHubsTargetScaler.cs

```diff
@@ -0,0 +1,186 @@
+
+                return previousResult;
+            }
+            else
+            {
+                return currentResult;
+            }
+        }
+
+        internal TargetScalerResult GetScaleResultInternal(TargetScalerContext context, long eventCount, int partitionCount)
+        {
+            int desiredConcurrency = GetDesiredConcurrencyInternal(context);
+            int desiredWorkerCount = (int)Math.Ceiling(eventCount / (decimal)desiredConcurrency);
+            int[] sortedValidWorkerCounts = GetSortedValidWorkerCountsForPartitionCount(partitionCount);
+            int validatedTargetWorkerCount = GetValidWorkerCount(desiredWorkerCount, sortedValidWorkerCounts);
+
+            _logger.LogInformation($"'Target worker count for function '{_functionId}' is '{validatedTargetWorkerCount}' (EventHubName='{_client.EventHubName}', EventCount ='{eventCount}', Concurrency='{desiredConcurrency}').");
+
+            return new TargetScalerResult
+            {
+                TargetWorkerCount = validatedTargetWorkerCount,
+            };
+        }
+
+        internal int GetDesiredConcurrencyInternal(TargetScalerContext context)
+        {
+            int concurrency = 0;
+
+            if (!context.InstanceConcurrency.HasValue)
+            {
+                if (_options.TargetUnprocessedEventThreshold > 0)
+                {
+                    concurrency = (int)_options.TargetUnprocessedEventThreshold;
+                }
+                else if (_options.MaxEventBatchSize > 0)
+                {
+                    concurrency = _options.MaxEventBatchSize;
+                }
+            }
+            else
+            {
+                concurrency = context.InstanceConcurrency.Value;
+            }
+
+            if (concurrency < 1)
+            {
+                throw new ArgumentOutOfRangeException($"Unexpected concurrency='{concurrency}'. Concurrency must be > 0.");
+            }
+            else
+            {
+                return concurrency;
+            }
+        }
+        /// <summary>
+        /// Returns an ordered list of valid worker counts for a given Event Hub partitionCount.
+        /// </summary>
+        /// <param name="partitionCount">Partition count of EventHub. Must be greater than 0.</param>
+        /// <returns></returns>
+        internal static int[] GetSortedValidWorkerCountsForPartitionCount(int partitionCount)
+        {
+            if (partitionCount == 0)
+            {
+                return Enumerable.Range(0, partitionCount).ToArray();
+            }
+            else
+            {
+                List<int> list = Enumerable.Range(1, partitionCount).
+                    Select(c => (double)partitionCount / (double)c).
+                    Select(d => (int)Math.Ceiling((double)d)).
+                    Distinct().
+                    OrderBy(e => e).
+                    ToList();
+                list.Insert(index: 0, item: 0);
```

Comment:
Seems odd to allocate a list just to do an insertion in the beginning and then to allocate an array after.   Maybe consider just creating the array upfront and then filling via `foreach` rather than LINQ?

Reply:
Resolving from discussion via offline conversation -- distinct can change length of array allocated in line 143.